### PR TITLE
1.0.0-alpha1

### PR DESCRIPTION
**Pull request by cdavernas**
**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts of this PR update:**

- [x] Specification
- [x] Schema
- [x] Examples
- [x] Extensions
- [x] Roadmap
- [x] Use Cases
- [x] Community
- [x] TCK
- [x] Other

**Discussion or Issue link**:
#843

**What this PR does / why we need it**:

The version `1.0.0-alpha1` is one of the most impacting release of the history of the Serverless Workflow DSL. It aims at both drastically simplify the DSL and at completing it with many new features.

It is intended to be the last release containing breaking changes before the Serverless Workflow DSL `1.0.0` can see the day.

## What's new?

- Total refactor of grammar and semantics. See [motivation](#). 
*Addresses #672.*
- Moved all descriptive properties into the document top level property. Adds a new (required) `namespace` DNS Label property. 
*Addresses #838.*
- Complex objects use maps instead of arrays, thus enabling more intelligible JSON pointers. 
*Addresses #831.*
- All workflow components are now referencable. This allows getting rid of many cumbersome 'oneOf' definitions. 
*Partially addresses #691.*
- Merged the `state` and `action` concepts into `task`, getting rid of state nesting limitations. See [tasks](#).
- Merged `end` and `transition` into [flow directives](#). 
*Addresses #684.*
- Got rid of the concept of compensation. Compensation can easily be designed by branching and/or chaining tasks.
- Support task nesting, thanks to [composite tasks](#). 
*Addresses #413.*
- Allows calling functions over HTTP, GRPC, OpenAPI and AsyncAPI. See [calls](#). 
*Addresses #828.*
- Allows running custom processes, using containers, shell commands, scripts or workflows. See [processes](#).
- Supports both sequential and concurrent task execution modes. See [execution strategy](#).
- Offers fault tolerance features, using [try task](#) to catch and handle [errors](#) raised during execution.
- Enables retrying on specific errors. See [retry policy](#). 
*Addresses #681 and #772.*
- Supports timeouts, used to define a duration after which to raise a new timeout error, possibly faulting the task/workflow. See [timeouts](#).
- Allows raising errors. See [errors](#errors). 
*Addresses #771.*
- Added a serie of Gherkin features used for both conformance testing and Behavior Driven Design (BDD). 
*Addresses #324.*
- Supports concurrent event correlations. 
*Addresses #679 and #680.*
- Enables extending and overriding the DSL, in a runtime agnostic and fully portable fashion, without writting a single line of code. See [extensions](#).

### CODE REVIEW

**Comment by cdavernas** [comment]
WIP

**Comment by cdavernas** [comment]
@ricardozanini @fjtirado I took the freedom of renaming `TCK`, which is apparently specifically used for Java, by `CTK`, but it does however change the meaning a bit from `Technology Compatibility Kit` into `Compliance Test Kit`. There is also the option of simply changing the meaning of the acronym into `Technology Compliance Kit` or something like that.

WDYT?

**Comment by cdavernas** [comment]
@ricardozanini @fjtirado I removed the examples instead of just updating them, because I believe this role is fullfilled by the Gherkin features, and is even more intelligible imho. What do you guys think?

**Comment by cdavernas** [comment]
@ricardozanini @fjtirado Could you guys also take care of creating use cases (for documentation), and Gherkin tests, if need be, for the CTK?

**Comment by ricardozanini** [comment]
> @ricardozanini Do we need to include the annual reviews, and if so, can we move them to a properly cased directory (such as `annual-reviews`)?

We can keep those in the folder `annual-reviews`, but I think CNCF is ditching those. Until we can confirm this info I think we can keep it there.

> What about the meetingminutes directory? Do we still need it? If so, can we properly case it (such as `meeting-minutes`, or move it to a relevant subdirectory such as `community`)

Please ditch this one, we have the discussions: https://github.com/serverlessworkflow/specification/discussions/categories/meetings-notes

**Comment by ricardozanini** [comment]
> @ricardozanini We will need to refactor comparisons, if we believe it's necessary to keep them. If so, could you guys @ricardozanini @fjtirado take care of it?

I don't think we need those.

**Comment by ricardozanini** [comment]
> @ricardozanini @fjtirado I removed the examples instead of just updating them, because I believe this role is fullfilled by the Gherkin features, and is even more intelligible imho. What do you guys think?

We need to be able to validate the schema somehow. The examples are good for this, plus newcomers can grasp the DSL easier with them, IMO.

**Comment by ricardozanini** [comment]
> @ricardozanini @fjtirado Could you guys also take care of creating use cases (for documentation), and Gherkin tests, if need be, for the CTK?

Sure thing, it will also be a great way to validate our refactoring and use cases.

**Comment by cdavernas** [comment]
> We need to be able to validate the schema somehow. The examples are good for this, plus newcomers can grasp the DSL easier with them, IMO.

Indeed. As for the schema, I'm working on it. It's a big piece.

**Comment by cdavernas** [comment]
@ricardozanini Should we keep splitting the schema accross multiple files? Or not? Or more than before? Or maybe less?

**Comment by JBBianchi** [comment]
> @ricardozanini @JBBianchi Should we keep splitting the schema accross multiple files? Or not? Or more than before? Or maybe less?

If a "resource" can be external to a workflow then I guess it can make sense to have an external schema to validate it. Otherwise, I don't really see the point but maybe I'm missing something.

**Comment by cdavernas** [comment]
@JBBianchi Most can be. That's not the question imo, as even like that it could be contained in a single file's $defs

File: dsl.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+- Utilize imperative verbs to convey actions directly and clearly.
+- Employ implicit default behaviors to reduce redundancy and streamline workflow definitions.
+- Enable the declaration and effortless import of shared components by supporting external references
+- Encourage the declaration of components inline for situations where reusability is unnecessary, prioritizing ease of use in such cases.
+- Prioritize flexibility over strong-typed enumerations for enhanced extensibility.
+- Opt for universally understood terms to improve accessibility and comprehension.
+
+## Versioning
+
+## Version
```
Comment (ricardozanini) [inline dsl.md:40]: I think we can move this to `CONTRIBUTIONS.md`. Might be confusing in the dsl docs.

**Review by ricardozanini — COMMENTED** [review]
Quick review to start our discussions.

**Comment by ricardozanini** [comment]
> @ricardozanini @JBBianchi Should we keep splitting the schema accross multiple files? Or not? Or more than before? Or maybe less?

We can keep it in one file, it's easier to manipulate it in a single file.

File: dsl-reference.md [hunk 2]
```diff
@@ -0,0 +1,1384 @@
...
+do:
+  runContainer:
+    run:
+      container:
+        image: fake-image
+```
+
+##### Script Process
+
+Enables the execution of custom scripts or code within a workflow, empowering workflows to perform specialized logic, data processing, or integration tasks by executing user-defined scripts written in various programming languages.
```
[the remark this answers is not in the document]
Comment (cdavernas) [reply dsl-reference.md:664]: I think we should impose `script` indeed, however probably only `javascript` and `python`, which are the most commonly used. WDYT?

File: dsl-reference.md [hunk 3]
```diff
@@ -0,0 +1,1384 @@
...
+    run:
+      script:
+        language: js
+        code: >
+          Some cool multiline script
+```
+
+##### Shell Process
+
+Enables the execution of shell commands within a workflow, enabling workflows to interact with the underlying operating system and perform system-level operations, such as file manipulation, environment configuration, or system administration tasks.
```
[the remark this answers is not in the document]
Comment (cdavernas) [reply dsl-reference.md:693]: I think here we should impose an `OS`. Hell, even microsoft recommends using Linux, when running containers. 

So yeah, I think we should document that it should be isolated within a Linux (Unix?) based environment.

WDYT?

File: dsl-reference.md [hunk 4]
```diff
@@ -0,0 +1,1384 @@
...
+| [https://Serverless Workflow.io/spec/1.0.0/errors/authentication](#) | `401` | Errors related to authentication failures. |
+| [https://Serverless Workflow.io/spec/1.0.0/errors/authorization](#) | `403` | Errors related to unauthorized access attempts or insufficient permissions to perform certain actions within the workflow. |
+| [https://Serverless Workflow.io/spec/1.0.0/errors/timeout](#) | `408` | Errors caused by timeouts during the execution of tasks or during interactions with external services. |
+| [https://Serverless Workflow.io/spec/1.0.0/errors/communication](#) | `500` | Errors  encountered while communicating with external services, including network errors, service unavailable, or invalid responses. |
+| [https://Serverless Workflow.io/spec/1.0.0/errors/runtime](#) | `500` | Errors occurring during the runtime execution of a workflow, including unexpected exceptions, errors related to resource allocation, or failures in handling workflow tasks. These errors typically occur during the actual execution of workflow components and may require runtime-specific handling and resolution strategies. |
```
[the remark this answers is not in the document]
Comment (cdavernas) [reply dsl-reference.md:1185]: 😆 Yeah, overlooked that, sorry!

File: dsl.md (hunk 1 above)
Comment (cdavernas) [reply dsl.md:40]: Allright, will do!

File: dsl-reference.md [hunk 5]
```diff
@@ -0,0 +1,1384 @@
...
+
+###### Properties
+
+| Name | Type | Required | Description|
+|:--|:---:|:---:|:---|
+| document | [`externalResource`](#external-resource) | `yes` | The OpenAPI document that defines the operation to call. |
+| operation | `string` | `yes` | The id of the OpenAPI operation to call. |
+| arguments | `map` | `no` | A name/value mapping of the parameters, if any, of the OpenAPI operation to call. |
+| authentication | [`authenticationPolicy`](#authentication-policy) | `no` | The authentication policy, or the name of the authentication policy, to use when calling the OpenAPI operation. |
+| output | `string` | `no` | The OpenAPI call's output format.<br>*Supported values are:*<br>*- `raw`, which output's the base-64 encoded [http response](#http-response) content, if any.*<br>*- `content`, which outputs the content of [http response](#http-response), possibly deserialized.*<br>*- `response`, which outputs the [http response](#http-response).*<br>*Defaults to `content`.* |
```
[the remark this answers is not in the document]
Comment (cdavernas) [reply dsl-reference.md:333]: No. `response ` is a wrapper around the deserialized raw. Raw is just the body of the http response, base-64 encoded.

[the remark this answers is not in the document]
Comment (cdavernas) [reply dsl-reference.md:333]: @ricardozanini Does that answer your question?

**Pull request closed without merging**